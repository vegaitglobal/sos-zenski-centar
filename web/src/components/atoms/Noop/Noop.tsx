import { ReactElement } from 'react';

export const Noop = (): ReactElement => <></>;
