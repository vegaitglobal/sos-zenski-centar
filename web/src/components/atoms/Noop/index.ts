export { Noop as default } from './Noop';
