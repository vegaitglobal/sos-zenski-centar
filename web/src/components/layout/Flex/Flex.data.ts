import { ElementType, ReactElement } from 'react';

export type FlexDirection = 'row' | 'column' | 'row-reverse' | 'column-reverse';
export type FlexJustify =
	| 'flex-start'
	| 'flex-end'
	| 'center'
	| 'space-between'
	| 'space-around'
	| 'initial'
	| 'inherit';
export type FlexAlign =
	| 'stretch'
	| 'center'
	| 'flex-start'
	| 'flex-end'
	| 'baseline'
	| 'initial'
	| 'inherit';
export type FlexWrap = 'wrap' | 'nowrap' | 'wrap-reverse';
export type FlexGrowShrink = '0' | '1';

export interface FlexProps {
	flexDirection?: FlexDirection;
	justifyContent?: FlexJustify;
	alignItems?: FlexAlign;
	flexWrap?: FlexWrap;
	flexGrow?: FlexGrowShrink;
	flexShrink?: FlexGrowShrink;
	elementType?: ElementType;
	children: ReactElement;
}

export interface StyledFlexProps {
	$flexDirection: FlexDirection;
	$justifyContent: FlexJustify;
	$alignItems: FlexAlign;
	$flexWrap: FlexWrap;
	$flexGrow: FlexGrowShrink;
	$flexShrink: FlexGrowShrink;
}
