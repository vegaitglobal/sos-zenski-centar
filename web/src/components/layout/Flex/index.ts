export { Flex as default } from './Flex';
