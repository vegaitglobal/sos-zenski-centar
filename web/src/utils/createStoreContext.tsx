import { createContext, PropsWithChildren, FunctionComponent, Context, useContext } from 'react';
import useStoreFactory from '../hooks/useStoreFactory';

/**
 * Creates a context for the store created by the passed `storeFactory`.
 * Additionally returns the `Provider` component and the hook for the created store's context.
 *
 * @param name - name of the store
 * @param storeFactory - a function that returns an object representing the store
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const createStoreContext = <TStore extends Record<string, unknown>>(
	name: string,
	storeFactory: () => TStore,
): [FunctionComponent<unknown>, () => TStore, Context<TStore>] => {
	const context = createContext<TStore>({} as TStore); // the initial value is only to satisfy TS and won't be used

	/**
	 * Initializes and contains the context for the `TStore`.
	 */
	function StoreProvider({ children }: PropsWithChildren<unknown>): JSX.Element {
		const store = useStoreFactory(storeFactory);
		return <context.Provider value={store}>{children}</context.Provider>;
	}

	const useStore = (): TStore => {
		const store = useContext(context);
		if (store) {
			return store;
		}
		throw new ReferenceError(
			`${name} is not initialized, make sure that your component is wrapped in the 'Provider' of the store.`,
		);
	};

	return [StoreProvider, useStore, context];
};

export default createStoreContext;
