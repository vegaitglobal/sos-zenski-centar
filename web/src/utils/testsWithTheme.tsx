import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { ReactElement, ComponentType, ReactNode } from 'react';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType = ({ children }) => componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
