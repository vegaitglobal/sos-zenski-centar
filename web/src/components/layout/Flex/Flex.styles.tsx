import styled from 'styled-components';
import { StyledFlexProps } from './Flex.data';

export const StyledFlex = styled.div<StyledFlexProps>`
	display: flex;
	${({ $flexDirection }) => $flexDirection && `flex-direction: ${$flexDirection};`}
	${({ $justifyContent }) => $justifyContent && `justify-content: ${$justifyContent};`}
	${({ $alignItems }) => $alignItems && `align-items: ${$alignItems};`}
	${({ $flexWrap }) => $flexWrap && `flex-wrap: ${$flexWrap};`}
  ${({ $flexGrow }) => $flexGrow && `flex-grow: ${$flexGrow};`}
  ${({ $flexShrink }) => $flexShrink && `flex-shrink: ${$flexShrink};`}
`;
