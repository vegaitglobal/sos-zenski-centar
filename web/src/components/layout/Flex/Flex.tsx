import { forwardRef } from 'react';
import { FlexProps } from './Flex.data';
import { StyledFlex } from './Flex.styles';

export const Flex = forwardRef<Element, FlexProps>(
	(
		{
			flexDirection,
			justifyContent,
			alignItems,
			flexWrap,
			flexGrow,
			flexShrink,
			children,
			elementType = 'div',
			...props
		},
		ref,
	) => (
		<StyledFlex
			data-testid="Flex"
			ref={ref}
			as={elementType}
			$flexDirection={flexDirection}
			$justifyContent={justifyContent}
			$alignItems={alignItems}
			$flexWrap={flexWrap}
			$flexGrow={flexGrow}
			$flexShrink={flexShrink}
			{...props}
		>
			{children}
		</StyledFlex>
	),
);
