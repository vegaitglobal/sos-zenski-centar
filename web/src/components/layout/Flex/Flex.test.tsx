import renderer from 'react-test-renderer';
import Flex from './index';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Noop from '../../atoms/Noop';

describe('Flex', () => {
	describe('Render', () => {
		it('Renders Flex', () => {
			const { getByTestId } = renderWithTheme(
				<Flex>
					<Noop />
				</Flex>,
			);

			const element = getByTestId('Flex');

			expect(element).toBeTruthy();
		});

		it('Renders header tag', () => {
			const { getByTestId } = renderWithTheme(
				<Flex elementType="header">
					<Noop />
				</Flex>,
			);

			const element = getByTestId('Flex');

			expect(element.tagName).toBe('HEADER');
		});

		it('Renders correctly', () => {
			const tree = renderer
				.create(
					<Flex>
						<Noop />
					</Flex>,
				)
				.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('Can have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Flex>
					<span>Child</span>
				</Flex>,
			);

			const element = getByTestId('Flex');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
