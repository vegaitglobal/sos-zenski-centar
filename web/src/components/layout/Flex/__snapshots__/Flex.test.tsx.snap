// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex Render Renders correctly 1`] = `
<div
  className="sc-bdnxRM dtpsiN"
  data-testid="Flex"
/>
`;
